File: packages_mjs/provider_api_getters/src/api_token/akamai.mjs
/**
 * Akamai Edge API token getter.
 *
 * This module provides API token resolution for Akamai Edge APIs.
 * Akamai uses EdgeGrid authentication, which requires:
 * - client_token
 * - client_secret
 * - access_token
 * - host
 *
 * These can be provided via environment variables or .edgerc file.
 *
 * API Documentation:
 *     https://techdocs.akamai.com/developer/docs/authenticate-with-edgegrid
 */
import { BaseApiToken, ApiKeyResult, maskSensitive } from './base.mjs';
import { existsSync, readFileSync } from 'fs';
import { homedir } from 'os';
import { join } from 'path';

// Simple logger
const logger = {
  debug: (msg) => console.debug(`[DEBUG] provider_api_getters.akamai: ${msg}`),
  warn: (msg) => console.warn(`[WARN] provider_api_getters.akamai: ${msg}`),
};

// Default .edgerc file location
const DEFAULT_EDGERC_PATH = join(homedir(), '.edgerc');

/**
 * EdgeGrid credentials container.
 */
class EdgeGridCredentials {
  constructor({ clientToken = null, clientSecret = null, accessToken = null, host = null } = {}) {
    this.clientToken = clientToken;
    this.clientSecret = clientSecret;
    this.accessToken = accessToken;
    this.host = host;
  }

  get isValid() {
    return !!(this.clientToken && this.clientSecret && this.accessToken && this.host);
  }
}

/**
 * Parse .edgerc INI file format.
 * @param {string} content - File content
 * @param {string} section - Section name to parse
 * @returns {Object} - Parsed key-value pairs
 */
function parseEdgerc(content, section = 'default') {
  const result = {};
  let currentSection = null;

  for (const line of content.split('\n')) {
    const trimmed = line.trim();

    // Skip empty lines and comments
    if (!trimmed || trimmed.startsWith('#') || trimmed.startsWith(';')) {
      continue;
    }

    // Check for section header
    const sectionMatch = trimmed.match(/^\[([^\]]+)\]$/);
    if (sectionMatch) {
      currentSection = sectionMatch[1];
      continue;
    }

    // Parse key = value
    if (currentSection === section) {
      const eqIndex = trimmed.indexOf('=');
      if (eqIndex !== -1) {
        const key = trimmed.slice(0, eqIndex).trim();
        const value = trimmed.slice(eqIndex + 1).trim();
        result[key] = value;
      }
    }
  }

  return result;
}

export class AkamaiApiToken extends BaseApiToken {
  get providerName() {
    return 'akamai';
  }

  get healthEndpoint() {
    logger.debug('AkamaiApiToken.healthEndpoint: Returning /-/client-api/active-grants/implicit');
    return '/-/client-api/active-grants/implicit';
  }

  /**
   * Get environment variable name from config for a specific key.
   * @param {string} key
   * @returns {string|null}
   */
  _getEnvVarName(key) {
    const providerConfig = this._getProviderConfig();
    if (providerConfig) {
      return providerConfig[`env_${key}`] || null;
    }
    return null;
  }

  /**
   * Get EdgeGrid credentials from environment variables.
   * @returns {EdgeGridCredentials}
   */
  _getCredentialsFromEnv() {
    logger.debug('AkamaiApiToken._getCredentialsFromEnv: Checking environment variables');

    const clientTokenVar = this._getEnvVarName('client_token') || 'AKAMAI_CLIENT_TOKEN';
    const clientSecretVar = this._getEnvVarName('client_secret') || 'AKAMAI_CLIENT_SECRET';
    const accessTokenVar = this._getEnvVarName('access_token') || 'AKAMAI_ACCESS_TOKEN';
    const hostVar = this._getEnvVarName('host') || 'AKAMAI_HOST';

    const credentials = new EdgeGridCredentials({
      clientToken: process.env[clientTokenVar],
      clientSecret: process.env[clientSecretVar],
      accessToken: process.env[accessTokenVar],
      host: process.env[hostVar],
    });

    if (credentials.isValid) {
      logger.debug('AkamaiApiToken._getCredentialsFromEnv: Found valid credentials in environment');
    } else {
      const missing = [];
      if (!credentials.clientToken) missing.push(clientTokenVar);
      if (!credentials.clientSecret) missing.push(clientSecretVar);
      if (!credentials.accessToken) missing.push(accessTokenVar);
      if (!credentials.host) missing.push(hostVar);
      logger.debug(`AkamaiApiToken._getCredentialsFromEnv: Missing env vars: [${missing.join(', ')}]`);
    }

    return credentials;
  }

  /**
   * Get EdgeGrid credentials from .edgerc file.
   * @param {string} section - Section name (default: 'default')
   * @returns {EdgeGridCredentials}
   */
  _getCredentialsFromEdgerc(section = 'default') {
    logger.debug(`AkamaiApiToken._getCredentialsFromEdgerc: Checking ${DEFAULT_EDGERC_PATH}`);

    if (!existsSync(DEFAULT_EDGERC_PATH)) {
      logger.debug(`AkamaiApiToken._getCredentialsFromEdgerc: File not found: ${DEFAULT_EDGERC_PATH}`);
      return new EdgeGridCredentials();
    }

    try {
      const content = readFileSync(DEFAULT_EDGERC_PATH, 'utf-8');
      const parsed = parseEdgerc(content, section);

      const credentials = new EdgeGridCredentials({
        clientToken: parsed.client_token || null,
        clientSecret: parsed.client_secret || null,
        accessToken: parsed.access_token || null,
        host: parsed.host || null,
      });

      if (credentials.isValid) {
        logger.debug(`AkamaiApiToken._getCredentialsFromEdgerc: Found valid credentials in [${section}]`);
      } else {
        logger.debug(`AkamaiApiToken._getCredentialsFromEdgerc: Incomplete credentials in [${section}]`);
      }

      return credentials;
    } catch (e) {
      logger.warn(`AkamaiApiToken._getCredentialsFromEdgerc: Error reading .edgerc: ${e.message}`);
      return new EdgeGridCredentials();
    }
  }

  /**
   * Get EdgeGrid credentials from environment or .edgerc file.
   * @returns {EdgeGridCredentials}
   */
  getCredentials() {
    // Try environment variables first
    let credentials = this._getCredentialsFromEnv();
    if (credentials.isValid) {
      return credentials;
    }

    // Fall back to .edgerc file
    credentials = this._getCredentialsFromEdgerc();
    return credentials;
  }

  getApiKey() {
    logger.debug('AkamaiApiToken.getApiKey: Starting credential resolution');

    const credentials = this.getCredentials();

    if (credentials.isValid) {
      // Store credentials as JSON in apiKey field
      const credentialsJson = JSON.stringify({
        client_token: credentials.clientToken,
        client_secret: credentials.clientSecret,
        access_token: credentials.accessToken,
        host: credentials.host,
      });

      logger.debug(
        `AkamaiApiToken.getApiKey: Found credentials for host ${maskSensitive(credentials.host || '')}`
      );

      const result = new ApiKeyResult({
        apiKey: credentialsJson,
        authType: 'edgegrid',
        headerName: 'Authorization',
        email: null,
        rawApiKey: credentials.accessToken,
        username: credentials.clientToken,
      });

      logger.debug(`AkamaiApiToken.getApiKey: Returning result hasCredentials=${result.hasCredentials}`);
      return result;
    } else {
      logger.warn(
        'AkamaiApiToken.getApiKey: No valid credentials found. ' +
        'Set AKAMAI_CLIENT_TOKEN, AKAMAI_CLIENT_SECRET, AKAMAI_ACCESS_TOKEN, ' +
        'AKAMAI_HOST environment variables or create ~/.edgerc file.'
      );

      const result = new ApiKeyResult({
        apiKey: null,
        authType: 'edgegrid',
        headerName: 'Authorization',
        email: null,
        rawApiKey: null,
      });

      logger.debug(`AkamaiApiToken.getApiKey: Returning result hasCredentials=${result.hasCredentials}`);
      return result;
    }
  }

  getBaseUrl() {
    const credentials = this.getCredentials();
    if (credentials.host) {
      // Ensure host has https:// prefix
      let host = credentials.host;
      if (!host.startsWith('https://')) {
        host = `https://${host}`;
      }
      return host;
    }
    return super.getBaseUrl();
  }
}

File: packages_py/provider_api_getters/src/provider_api_getters/api_token/akamai.py
"""
Akamai Edge API token getter.

This module provides API token resolution for Akamai Edge APIs.
Akamai uses EdgeGrid authentication, which requires:
- client_token
- client_secret
- access_token
- host

These can be provided via environment variables or .edgerc file.

API Documentation:
    https://techdocs.akamai.com/developer/docs/authenticate-with-edgegrid
"""
import logging
import os
from dataclasses import dataclass
from pathlib import Path
from typing import Optional, Tuple

from .base import BaseApiToken, ApiKeyResult, _mask_sensitive

logger = logging.getLogger(__name__)

# Default .edgerc file location
DEFAULT_EDGERC_PATH = Path.home() / ".edgerc"


@dataclass
class EdgeGridCredentials:
    """Akamai EdgeGrid credentials."""
    client_token: Optional[str] = None
    client_secret: Optional[str] = None
    access_token: Optional[str] = None
    host: Optional[str] = None

    @property
    def is_valid(self) -> bool:
        """Check if all required credentials are present."""
        return all([
            self.client_token,
            self.client_secret,
            self.access_token,
            self.host
        ])


class AkamaiApiToken(BaseApiToken):
    """
    API token getter for Akamai Edge APIs.

    Akamai uses EdgeGrid authentication which requires four credentials:
    - client_token: Client token from Akamai Control Center
    - client_secret: Client secret from Akamai Control Center
    - access_token: Access token from Akamai Control Center
    - host: Akamai API hostname (e.g., akab-xxx.luna.akamaiapis.net)

    Configuration:
        providers.akamai.base_url: null (derived from host)
        providers.akamai.env_client_token: "AKAMAI_CLIENT_TOKEN"
        providers.akamai.env_client_secret: "AKAMAI_CLIENT_SECRET"
        providers.akamai.env_access_token: "AKAMAI_ACCESS_TOKEN"
        providers.akamai.env_host: "AKAMAI_HOST"

    Environment Variables:
        AKAMAI_CLIENT_TOKEN: Client token
        AKAMAI_CLIENT_SECRET: Client secret
        AKAMAI_ACCESS_TOKEN: Access token
        AKAMAI_HOST: API hostname

    Alternative: .edgerc file
        If environment variables are not set, credentials can be read from
        ~/.edgerc file in INI format (default section).
    """

    @property
    def provider_name(self) -> str:
        """Return the provider name for Akamai."""
        return "akamai"

    @property
    def health_endpoint(self) -> str:
        """
        Return the health check endpoint for Akamai.

        The /-/client-api/active-grants/implicit endpoint returns
        information about the current API client's grants.
        """
        logger.debug("AkamaiApiToken.health_endpoint: Returning /-/client-api/active-grants/implicit")
        return "/-/client-api/active-grants/implicit"

    def _get_env_var_name(self, key: str) -> Optional[str]:
        """Get environment variable name from config for a specific key."""
        provider_config = self._get_provider_config()
        if provider_config:
            return provider_config.get(f"env_{key}")
        return None

    def _get_credentials_from_env(self) -> EdgeGridCredentials:
        """Get EdgeGrid credentials from environment variables."""
        logger.debug("AkamaiApiToken._get_credentials_from_env: Checking environment variables")

        client_token_var = self._get_env_var_name("client_token") or "AKAMAI_CLIENT_TOKEN"
        client_secret_var = self._get_env_var_name("client_secret") or "AKAMAI_CLIENT_SECRET"
        access_token_var = self._get_env_var_name("access_token") or "AKAMAI_ACCESS_TOKEN"
        host_var = self._get_env_var_name("host") or "AKAMAI_HOST"

        credentials = EdgeGridCredentials(
            client_token=os.getenv(client_token_var),
            client_secret=os.getenv(client_secret_var),
            access_token=os.getenv(access_token_var),
            host=os.getenv(host_var),
        )

        if credentials.is_valid:
            logger.debug("AkamaiApiToken._get_credentials_from_env: Found valid credentials in environment")
        else:
            missing = []
            if not credentials.client_token:
                missing.append(client_token_var)
            if not credentials.client_secret:
                missing.append(client_secret_var)
            if not credentials.access_token:
                missing.append(access_token_var)
            if not credentials.host:
                missing.append(host_var)
            logger.debug(f"AkamaiApiToken._get_credentials_from_env: Missing env vars: {missing}")

        return credentials

    def _get_credentials_from_edgerc(self, section: str = "default") -> EdgeGridCredentials:
        """
        Get EdgeGrid credentials from .edgerc file.

        The .edgerc file is in INI format:
            [default]
            client_secret = xxxx
            host = akab-xxx.luna.akamaiapis.net
            access_token = akab-xxx
            client_token = akab-xxx
        """
        logger.debug(f"AkamaiApiToken._get_credentials_from_edgerc: Checking {DEFAULT_EDGERC_PATH}")

        if not DEFAULT_EDGERC_PATH.exists():
            logger.debug(f"AkamaiApiToken._get_credentials_from_edgerc: File not found: {DEFAULT_EDGERC_PATH}")
            return EdgeGridCredentials()

        try:
            import configparser
            config = configparser.ConfigParser()
            config.read(DEFAULT_EDGERC_PATH)

            if section not in config:
                logger.debug(f"AkamaiApiToken._get_credentials_from_edgerc: Section [{section}] not found")
                return EdgeGridCredentials()

            credentials = EdgeGridCredentials(
                client_token=config.get(section, "client_token", fallback=None),
                client_secret=config.get(section, "client_secret", fallback=None),
                access_token=config.get(section, "access_token", fallback=None),
                host=config.get(section, "host", fallback=None),
            )

            if credentials.is_valid:
                logger.debug(f"AkamaiApiToken._get_credentials_from_edgerc: Found valid credentials in [{section}]")
            else:
                logger.debug(f"AkamaiApiToken._get_credentials_from_edgerc: Incomplete credentials in [{section}]")

            return credentials
        except Exception as e:
            logger.warning(f"AkamaiApiToken._get_credentials_from_edgerc: Error reading .edgerc: {e}")
            return EdgeGridCredentials()

    def get_credentials(self) -> EdgeGridCredentials:
        """
        Get EdgeGrid credentials from environment or .edgerc file.

        Priority:
        1. Environment variables
        2. .edgerc file [default] section
        """
        # Try environment variables first
        credentials = self._get_credentials_from_env()
        if credentials.is_valid:
            return credentials

        # Fall back to .edgerc file
        credentials = self._get_credentials_from_edgerc()
        return credentials

    def get_api_key(self) -> ApiKeyResult:
        """
        Get Akamai EdgeGrid credentials.

        For Akamai, the api_key field contains a JSON-encoded string with
        all four EdgeGrid credentials. The auth_type is "edgegrid".

        Returns:
            ApiKeyResult with EdgeGrid credentials
        """
        logger.debug("AkamaiApiToken.get_api_key: Starting credential resolution")

        credentials = self.get_credentials()

        if credentials.is_valid:
            import json
            # Store credentials as JSON in api_key field
            credentials_json = json.dumps({
                "client_token": credentials.client_token,
                "client_secret": credentials.client_secret,
                "access_token": credentials.access_token,
                "host": credentials.host,
            })

            logger.debug(
                f"AkamaiApiToken.get_api_key: Found credentials for host "
                f"{_mask_sensitive(credentials.host or '')}"
            )

            result = ApiKeyResult(
                api_key=credentials_json,
                auth_type="edgegrid",
                header_name="Authorization",  # EdgeGrid uses Authorization header
                email=None,
                raw_api_key=credentials.access_token,  # Use access_token as raw key
                username=credentials.client_token,  # Store client_token for reference
            )
        else:
            logger.warning(
                "AkamaiApiToken.get_api_key: No valid credentials found. "
                "Set AKAMAI_CLIENT_TOKEN, AKAMAI_CLIENT_SECRET, AKAMAI_ACCESS_TOKEN, "
                "AKAMAI_HOST environment variables or create ~/.edgerc file."
            )
            result = ApiKeyResult(
                api_key=None,
                auth_type="edgegrid",
                header_name="Authorization",
                email=None,
                raw_api_key=None,
            )

        logger.debug(
            f"AkamaiApiToken.get_api_key: Returning result "
            f"has_credentials={result.has_credentials}"
        )
        return result

    def get_base_url(self) -> Optional[str]:
        """
        Get the base URL for Akamai API.

        The base URL is derived from the host credential.
        """
        credentials = self.get_credentials()
        if credentials.host:
            # Ensure host has https:// prefix
            host = credentials.host
            if not host.startswith("https://"):
                host = f"https://{host}"
            return host
        return super().get_base_url()
